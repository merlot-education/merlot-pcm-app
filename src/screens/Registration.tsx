import React, { useCallback, useEffect, useState } from 'react'
import { Alert, Keyboard, ScrollView, StyleSheet, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import { useTranslation } from 'react-i18next'
import Toast from 'react-native-toast-message'
import { StackScreenProps } from '@react-navigation/stack'
import Clipboard from '@react-native-clipboard/clipboard'
import { setValueKeychain } from '../utils/keychain'
import { ColorPallet, TextTheme } from '../theme/theme'
import { TextInput, Loader, Text } from '../components'
import Button, { ButtonType } from '../components/button/Button'
import { OnboardingStackParams, Screens } from '../types/navigators'
import * as api from '../api'
import { ToastType } from '../components/toast/BaseToast'
import wordsList from '../utils/wordsList'
import { KeychainStorageKeys } from '../constants'

type RegistrationProps = StackScreenProps<
  OnboardingStackParams,
  Screens.Registration
>

const style = StyleSheet.create({
  container: {
    backgroundColor: ColorPallet.grayscale.white,
    margin: 20,
  },
  bodyText: {
    ...TextTheme.caption,
    flexShrink: 1,
    fontWeight: 'bold',
    marginVertical: 20,
  },
  boxContainer: {
    backgroundColor: ColorPallet.notification.info,
    borderRadius: 5,
    borderWidth: 1,
    borderColor: ColorPallet.notification.infoBorder,
    padding: 10,
    marginTop: 10,
    marginBottom: 20,
  },
  label: {
    ...TextTheme.normal,
    fontWeight: 'bold',
  },
  headerText: {
    ...TextTheme.normal,
    color: ColorPallet.notification.infoText,
    flexShrink: 1,
  },
})

const Registration: React.FC<RegistrationProps> = ({ navigation, route }) => {
  const { forgotPin } = route.params
  const [email, setEmail] = useState('')
  const [loading, setLoading] = useState(false)
  const [mnemonicText, setMnemonicText] = useState('')

  const { t } = useTranslation()

  const createMnemonic = useCallback(() => {
    const wordsArray = []
    let mnemonic = ''
    for (let index = 1; index <= 8; index += 1) {
      let diceNumber = ''
      for (let mnemonicWord = 0; mnemonicWord < 5; mnemonicWord += 1) {
        const num = Math.floor(Math.random() * 6) + 1
        diceNumber += num
      }
      const element = wordsList[diceNumber]
      wordsArray.push(element)
      mnemonic += `${element} `
    }
    setMnemonicText(mnemonic)
  }, [])

  useEffect(() => {
    createMnemonic()
  }, [createMnemonic])

  const emailCreate = async (emailId: string) => {
    const description = t('Registration.UserAuthenticationEmail')
    try {
      setValueKeychain(description, emailId, {
        service: 'email',
      })
    } catch (e) {
      Alert.alert(e)
    }
  }

  const savePassphrase = async (passphrase: string) => {
    const description = t('Registration.Passphrase')
    try {
      setValueKeychain(description, passphrase, {
        service: KeychainStorageKeys.Passphrase,
      })
    } catch (e) {
      Alert.alert(e)
    }
  }

  const validateEmail = email => {
    return email.match(
      // eslint-disable-next-line no-control-regex
      /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/,
    )
  }

  const confirmEntry = async (email: string) => {
    if (email.length > 0) {
      if (validateEmail(email)) {
        await emailCreate(email)
        if (!forgotPin) {
          await savePassphrase(mnemonicText)
        }
        try {
          setLoading(true)
          const {
            data: { otpId },
            message,
          } = await api.default.auth.register({
            email,
            otpId: '',
          })
          setLoading(false)
          Toast.show({
            type: ToastType.Success,
            text1: t('Toasts.Success'),
            text2: message,
          })
          if (forgotPin) {
            navigation.navigate(Screens.VerifyOtp, {
              email,
              forgotPin,
              otpId,
            })
          } else {
            navigation.navigate(Screens.VerifyOtp, {
              email,
              forgotPin,
              otpId,
            })
          }
        } catch (error) {
          setLoading(false)
          Toast.show({
            type: ToastType.Error,
            text1: error.name,
            text2: error.message,
          })
        }
      } else {
        Toast.show({
          type: ToastType.Warn,
          text1: t('Toasts.Warning'),
          text2: t('Registration.ValidEmail'),
        })
      }
    } else {
      Toast.show({
        type: ToastType.Warn,
        text1: t('Toasts.Warning'),
        text2: t('Registration.EnterEmail'),
      })
    }
    // navigation.navigate(Screens.CreatePin, { forgotPin })
  }

  return (
    <SafeAreaView style={style.container}>
      <ScrollView showsVerticalScrollIndicator={false}>
        <Loader loading={loading} />
        <TextInput
          label={t('Global.Email')}
          placeholder={t('Global.Email')}
          placeholderTextColor={ColorPallet.brand.primary}
          accessible
          accessibilityLabel={t('Global.Email')}
          autoFocus
          value={email}
          onChangeText={setEmail}
          keyboardType="email-address"
          autoCapitalize="none"
        />
        {!forgotPin && (
          <>
            <Text style={style.label}>Mnemonic</Text>
            <View style={style.boxContainer}>
              <Text style={style.headerText}>{mnemonicText}</Text>
              <Text style={style.bodyText}>
                {t('Registration.MnemonicMsg')}
              </Text>
              <Button
                title={t('Global.Copy')}
                buttonType={ButtonType.Primary}
                onPress={() => Clipboard.setString(mnemonicText)}
              />
            </View>
          </>
        )}
        <Button
          title={t('Global.Submit')}
          buttonType={ButtonType.Primary}
          onPress={() => {
            Keyboard.dismiss()
            confirmEntry(email)
          }}
        />
      </ScrollView>
    </SafeAreaView>
  )
}

export default Registration
