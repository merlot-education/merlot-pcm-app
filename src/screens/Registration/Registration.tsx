import React, { useCallback, useState } from 'react'
import { BackHandler, StyleSheet, View, Image, Keyboard } from 'react-native'
import { useTranslation } from 'react-i18next'
import { StackScreenProps } from '@react-navigation/stack'
import { useFocusEffect, useNavigation } from '@react-navigation/core'
import { ColorPallet, TextTheme } from '../../theme/theme'
import {
  TextInput,
  Loader,
  InfoCard,
  ScreenNavigatorButtons,
} from '../../components'
import { OnboardingStackParams, Screens } from '../../types/navigators'
import { KeychainStorageKeys } from '../../constants'
import {
  registerUser,
  restoreTermsCompleteStage,
  saveValueInKeychain,
  validateEmail,
} from './Registration.utils'
import Images from '../../assets'

type RegistrationProps = StackScreenProps<
  OnboardingStackParams,
  Screens.Registration
>

const style = StyleSheet.create({
  container: {
    backgroundColor: ColorPallet.grayscale.white,
    flex: 1,
    justifyContent: 'space-between',
    margin: 20,
  },

  subContainer: {
    backgroundColor: ColorPallet.grayscale.white,
    flex: 1,
    margin: 20,
  },
  bodyText: {
    ...TextTheme.caption,
    flexShrink: 1,
    fontWeight: 'bold',
    marginVertical: 20,
  },
  boxContainer: {
    backgroundColor: ColorPallet.notification.info,
    borderRadius: 5,
    borderWidth: 1,
    borderColor: ColorPallet.notification.infoBorder,
    padding: 10,
    marginTop: 10,
    marginBottom: 20,
  },
  emailImg: {
    height: 120,
    width: 60,
    alignSelf: 'center',
  },
  label: {
    ...TextTheme.normal,
    fontWeight: 'bold',
  },
  headerText: {
    ...TextTheme.normal,
    color: ColorPallet.notification.infoText,
    flexShrink: 1,
  },
})

const Registration: React.FC<RegistrationProps> = ({ navigation }) => {
  const [email, setEmail] = useState('')
  const [loading, setLoading] = useState(false)
  const [error, setError] = useState('')
  const nav = useNavigation()
  const { t } = useTranslation()

  const confirmEntry = async (email: string) => {
    if (email.length > 0) {
      if (validateEmail(email)) {
        await saveValueInKeychain(
          KeychainStorageKeys.Email,
          email,
          t('Registration.UserAuthenticationEmail'),
        )
        try {
          setLoading(true)
          const {
            data: { otpId },
            message,
          } = await registerUser(email, '')
          setLoading(false)
          // successToast(message)
          setError(message)
          navigation.navigate(Screens.VerifyOtp, { email, otpId })
        } catch (error) {
          setLoading(false)
          // errorToast(error.message)
          setError(error.message)
        }
      } else {
        setError(t('Registration.ValidEmail'))
      }
    } else {
      setError(t('Registration.EnterEmail'))
    }
  }

  const onBack = async () => {
    nav.navigate(Screens.Terms)
  }

  useFocusEffect(
    useCallback(() => {
      const onBackPress = async () => {
        await restoreTermsCompleteStage()
        nav.navigate(Screens.Terms)
        return true
      }

      BackHandler.addEventListener('hardwareBackPress', onBackPress)

      return () =>
        BackHandler.removeEventListener('hardwareBackPress', onBackPress)
    }, [nav]),
  )

  return (
    <View style={style.container}>
      <Loader loading={loading} />
      <TextInput
        label={t('Global.Email')}
        placeholder={t('Global.Email')}
        placeholderTextColor={ColorPallet.brand.primary}
        accessible
        accessibilityLabel={t('Global.Email')}
        autoFocus
        value={email}
        onChangeText={setEmail}
        keyboardType="email-address"
        autoCapitalize="none"
        returnKeyType="done"
      />
      <Image source={Images.emailIcon} style={style.emailImg} />
      <InfoCard showBottomIcon={false} showTopIcon errorMsg={error}>
        {t('Registration.EmailInfo')}
      </InfoCard>
      <ScreenNavigatorButtons
        onLeftPress={onBack}
        onRightPress={() => {
          Keyboard.dismiss()
          confirmEntry(email)
        }}
      />
    </View>
  )
}

export default Registration
