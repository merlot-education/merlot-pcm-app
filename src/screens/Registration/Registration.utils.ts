import AsyncStorage from '@react-native-async-storage/async-storage'
import { Alert } from 'react-native'
import api from '../../api'
import { EMAIL_REGEX, LocalStorageKeys } from '../../constants'
import { setValueKeychain } from '../../utils/keychain'
import wordsList from '../../utils/wordsList'

export const getMnemonicArrayFromWords = (lengthOfWords: number): string[] => {
  const wordsArray = []
  for (let index = 1; index <= lengthOfWords; index += 1) {
    let diceNumber = ''
    for (let mnemonicWord = 0; mnemonicWord < 5; mnemonicWord += 1) {
      const num = Math.floor(Math.random() * 6) + 1
      diceNumber += num
    }
    const element = wordsList[diceNumber]
    wordsArray.push(element)
  }
  return wordsArray
}

export const saveValueInKeychain = async (
  service: string,
  value: string,
  description: string,
) => {
  try {
    await setValueKeychain(description, value, {
      service,
    })
  } catch (e) {
    Alert.alert(e)
  }
}

export const registerUser = async (email: string, otpId: string) => {
  const data = await api.auth.register({
    email,
    otpId,
  })
  return data
}

export const validateEmail = (email: string): boolean => {
  return EMAIL_REGEX.test(email)
}

export const restoreTermsCompleteStage = async () => {
  await AsyncStorage.removeItem(LocalStorageKeys.OnboardingCompleteStage)
}
