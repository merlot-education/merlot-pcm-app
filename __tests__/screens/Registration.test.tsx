import React from 'react'
import { create } from 'react-test-renderer'
import Registration from '../../src/screens/Registration'

jest.mock('react-native-keychain', () => 'Registration')
jest.mock('react-native-config', () => 'Registration')

describe('Registration Screen', () => {
  it('Renders correctly', () => {
    const tree = create(<Registration />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
