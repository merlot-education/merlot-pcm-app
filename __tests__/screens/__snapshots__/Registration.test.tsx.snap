// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration Screen Renders correctly 1`] = `
<RNCSafeAreaView
  style={
    Array [
      Object {
        "backgroundColor": "#FFF",
        "margin": 20,
      },
    ]
  }
>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(52, 52, 52, 0.8)",
          "flex": 1,
          "justifyContent": "space-around",
        }
      }
    >
      <ActivityIndicator
        color="#03028A"
        size="small"
      />
    </View>
  </Modal>
  <View
    style={
      Object {
        "marginVertical": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#03028A",
          "fontSize": 14,
          "fontWeight": "bold",
          "marginBottom": 3,
        }
      }
    >
      Global.Email
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderBottomColor": "#dddddd",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "height": 44.5,
            "marginBottom": 0,
            "marginLeft": 15,
            "marginTop": 0,
            "paddingRight": 15,
          },
          Object {
            "borderBottomWidth": 0.5,
          },
        ]
      }
    >
      <TextInput
        accessibilityLabel="Global.Email"
        accessible={true}
        autoCapitalize="none"
        autoFocus={true}
        clearButtonMode="never"
        editable={true}
        keyboardType="email-address"
        labelPosition="left"
        onBlur={[Function]}
        onChange={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Global.Email"
        placeholderTextColor="#03028A"
        secureTextEntry={false}
        selectionColor="#03028A"
        style={
          Array [
            Object {
              "height": 35,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#000000",
              "flex": 1,
              "fontSize": 17,
            },
            null,
            Object {},
            Array [
              Object {
                "backgroundColor": "#FFF",
                "borderRadius": 4,
                "borderWidth": 2,
                "color": "#03028A",
                "fontSize": 16,
                "padding": 10,
              },
              false,
            ],
          ]
        }
        textAlign="left"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 5,
          "borderWidth": 1,
          "justifyContent": "center",
        },
        Object {
          "height": 47,
          "paddingLeft": 15,
          "paddingRight": 15,
        },
        Object {
          "backgroundColor": "#108ee9",
          "borderColor": "#108ee9",
        },
        false,
        false,
        undefined,
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 18,
            },
            Object {
              "color": "#ffffff",
            },
            false,
            false,
          ]
        }
      >
        Global.Submit
      </Text>
    </View>
  </View>
</RNCSafeAreaView>
`;
